File: algorithms/code/leetcode/lc114_flatten_binary_tree_to_linked_list/lc114_flatten_binary_tree_to_linked_list.py
#!/usr/bin/env python3
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: 'TreeNode | None' = None,
        right: 'TreeNode | None' = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flatten(self, root: TreeNode | None) -> None:
        while root is not None:
            if root.left is not None:
                node = root.left
                while node.right is not None:
                    node = node.right
                node.right = root.right
                root.right = root.left
                root.left = None
            root = root.right


class TestCode(unittest.TestCase):
    def __test(self, expected: TreeNode | None, root: TreeNode | None) -> None:
        if None is expected:
            self.assertEqual(None, root)
        else:
            self.assertIsNotNone(root)
            assert root is not None and expected is not None
            self.assertEqual(expected.val, root.val)
            self.__test(expected.left, root.left)
            self.__test(expected.right, root.right)

    def test_example(self) -> None:
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.left.left = TreeNode(3)
        root.left.right = TreeNode(4)
        root.right = TreeNode(5)
        root.right.right = TreeNode(6)
        expected = TreeNode(1)
        expected.right = TreeNode(2)
        expected.right.right = TreeNode(3)
        expected.right.right.right = TreeNode(4)
        expected.right.right.right.right = TreeNode(5)
        expected.right.right.right.right.right = TreeNode(6)
        Solution().flatten(root)
        self.__test(expected, root)

    def test_other_1(self) -> None:
        root = TreeNode(1)
        root.right = TreeNode(2)
        root.right.left = TreeNode(3)
        expected = TreeNode(1)
        expected.right = TreeNode(2)
        expected.right.right = TreeNode(3)
        Solution().flatten(root)
        self.__test(expected, root)

    def test_other_2(self) -> None:
        root = TreeNode(3)
        root.left = TreeNode(1)
        root.left.left = TreeNode(4)
        root.left.left.right = TreeNode(2)
        expected = TreeNode(3)
        expected.right = TreeNode(1)
        expected.right.right = TreeNode(4)
        expected.right.right.right = TreeNode(2)
        Solution().flatten(root)
        self.__test(expected, root)

File: algorithms/code/leetcode/lc114_flatten_binary_tree_to_linked_list/lc114_flatten_binary_tree_to_linked_list.rs
// https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
// #medium

use std::cell::RefCell;
use std::rc::Rc;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    pub fn flatten(root: &mut Option<Rc<RefCell<TreeNode>>>) {
        let mut root = (*root).clone();
        while let Some(some_root) = root {
            let mut some_root_borrowed = some_root.borrow_mut();
            if let Some(some_root_left) = some_root_borrowed.left.clone() {
                let mut node = some_root_left.clone();
                let mut node_right = node.borrow().right.clone();
                while let Some(some_node_right) = node_right {
                    node = some_node_right.clone();
                    node_right = some_node_right.borrow().right.clone();
                }
                node.borrow_mut().right = some_root_borrowed.right.clone();
                some_root_borrowed.right = some_root_borrowed.left.take();
            }
            root = some_root_borrowed.right.clone();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test(expected: Option<Rc<RefCell<TreeNode>>>, result: Option<Rc<RefCell<TreeNode>>>) {
        assert!(!(expected.is_none() && result.is_some()));
        assert!(!(expected.is_some() && result.is_none()));
        if let (Some(some_expected), Some(some_result)) = (expected.clone(), result.clone()) {
            assert_eq!(some_expected.borrow().val, some_result.borrow().val);
            test(
                some_expected.borrow().left.clone(),
                some_result.borrow().left.clone(),
            );
            test(
                some_expected.borrow().left.clone(),
                some_result.borrow().left.clone(),
            );
        }
    }

    #[test]
    fn test_example() {
        let n6 = TreeNode::new(6, None, None);
        let n5 = TreeNode::new(5, None, n6);
        let n4 = TreeNode::new(4, None, None);
        let n3 = TreeNode::new(3, None, None);
        let n2 = TreeNode::new(2, n3, n4);
        let mut root = TreeNode::new(1, n2, n5);
        let e6 = TreeNode::new(6, None, None);
        let e5 = TreeNode::new(5, None, e6);
        let e4 = TreeNode::new(4, None, e5);
        let e3 = TreeNode::new(3, None, e4);
        let e2 = TreeNode::new(2, None, e3);
        let expected = TreeNode::new(1, None, e2);
        Solution::flatten(&mut root);
        test(expected, root);
    }

    #[test]
    fn test_other_1() {
        let n3 = TreeNode::new(3, None, None);
        let n2 = TreeNode::new(2, n3, None);
        let mut root = TreeNode::new(1, None, n2);
        let e3 = TreeNode::new(3, None, None);
        let e2 = TreeNode::new(2, None, e3);
        let expected = TreeNode::new(1, None, e2);
        Solution::flatten(&mut root);
        test(expected, root);
    }

    #[test]
    fn test_other_2() {
        let n2 = TreeNode::new(2, None, None);
        let n4 = TreeNode::new(4, n2, None);
        let n1 = TreeNode::new(1, n4, None);
        let mut root = TreeNode::new(3, n1, None);
        let e2 = TreeNode::new(2, None, None);
        let e4 = TreeNode::new(4, None, e2);
        let e1 = TreeNode::new(1, None, e4);
        let expected = TreeNode::new(3, None, e1);
        Solution::flatten(&mut root);
        test(expected, root);
    }
}
